File: problems/cpp/plus-one.cc
// 3 ms, 09/09/2017
// Time  :
// Space :
class Solution {
public:
    vector<int> plusOne(vector<int>& digits)
    {
        int carry = 1;
        for (int n = digits.size(), i = n - 1; i >= 0; --i) {
            digits[i] += carry;
            carry = digits[i] / 10;
            digits[i] %= 10;
        }
        if (carry)
            digits.insert(digits.begin(), carry);
        return digits;
    }
};

File: code/python/plus-one.py
# 01/12/2018, 63 ms
# Time  : O(n)
# Space : O(1)
class Solution:
    def plusOne(self, digits):
        """
        :type digits: List[int]
        :rtype: List[int]
        """
        carry = 1
        for i, d in reversed(list(enumerate(digits))):
            if not carry:
                break
            if d == 9:
                digits[i], carry = 0, 1
            else:
                digits[i], carry = d + 1, 0
        return [1] + digits if carry else digits
